generic container for jwt claims and vp tokens (#26)
awoie/vp-token-spec: README.md

There are two credential formats to VCs and VPs: JSON or JSON-LD. There are also two proof formats to VCs and VPs: JWT and Linked Data Proofs. Each of those formats has different properties and capabilites and each of them comes with different proof types. Proof formats are agnostic to the credential format chosen. However, the JSON credential format is commonly used with JSON Web Signatures (https://www.w3.org/TR/vc-data-model/#json-web-token). JSON-LD is commonly used with different kinds of Linked Data Proofs and JSON Web Signatures (https://www.w3.org/TR/vc-data-model/#json-ld). 

> This specification introduces the following mechanisms to exchange verifiable credentials objectes between OpenID OPs and RPs.

Reviewer: ```suggestion
This specification introduces the following mechanisms to exchange verifiable credentials objectes between OpenID OPs and RPs.
```

* JWT claims (`verifiable_presentations` and `verifiable_credentials`) used as genetic container objects to embed verifieble credential objects into ID tokens or userinfo responses.

> * New tokens types VP Token and VC Token contain verifiable credentials objectes

Reviewer 2: I would propose leave out VC token and anything related to VC out of the text as suggested during the last SIOP special call. Putting VC directly into the ID Token might lead to ID Token being treated as a new wrapper for VC competing with VP, and might allow OP to return a VC without a proof that it is a subject to whom VC was issued actually presenting it.

But it may also be a good idea to keep this option and take it out if it will not be used.

Author: will remove vc stuff

> * JWT claims (`verifiable_presentations` and `verifiable_credentials`) used as genetic container objects to embed verifieble credential objects into ID tokens or userinfo responses.
> * New tokens types VP Token and VC Token contain verifiable credentials objectes

Reviewer 3: ```suggestion
* JWT claims (`verifiable_presentations` and `verifiable_credentials`) used as generic container objects to embed verifiable credential objects into ID tokens or userinfo responses.
* New token types VP Token and VC Token contain verifiable credentials objects.
```

This specification defines the JWT parameters to pass VPs and VCs signed as JWTs or using Linked Data Proofs.

This specifications also introduces a new artifact VP Token and VC Token to provide VPs and VCs to RPs.

All representations share the same container format.

## Container Format

### Verifiable Credential

A verifiable credential container is an array of objects, each of them containing the following fields:

> `format`: REQUIRED A JSON string denoting the proof format the credential was returned in. This specification introduces the values `w3cvc-jsonld` and `w3cvc-jwt` to denote credentials in JSON-LD and JWT format, respectively.  

Reviewer 2: Usually VP in JSON format is signed as a JWT, but technically it can also be signed as LD-Proof. For VP in JSON-LD, (correct me if I am wrong) the only meaningful option is signing using LD-Proofs.

Probably we need both fields credential_format - JSON and JSON-LD and proof_format - JWT or LD-Proof? If not, it should be stated that w3cvc-jwt format includes VC in JSON and w3cvc-jsonld includes VC in JSON-LD.

This confusion between credential format and its proof format has to be distinguished...

Author: @awoie @AdamJLemmon @tplooker what do you think?

Reviewer 3: I would use the format designators from DIF PE [here](https://identity.foundation/presentation-exchange/#claim-format-designations) which are: jwt_vc, jwt_vp, ldp_vc, ldp_vp. It also defines jwt and ldp but I don't know how they are used.

I agree with @Sakurann that JSON-LD is not a proof type and I agree that the VC can be interpreted as plain JSON since the [VC spec says](https://www.w3.org/TR/vc-data-model/#contexts):
> Though this specification requires that a @context property be present, it is not required that the value of the @context property be processed using JSON-LD. This is to support processing using plain JSON libraries, such as those that might be used when the verifiable credential is encoded as a JWT. All libraries or processors MUST ensure that the order of the values in the @context property is what is expected for the specific application. Libraries or processors that support JSON-LD can process the @context property using full JSON-LD processing as expected.

However, since in practice, each VC should have a valid `@context` property and valid URIs, therefore being valid JSON-LD, I'm not sure if this is really required. If you feel it makes a difference, I'm ok with adding support for that case.

Author: changed identifiers to follow DIF PE definition

Reviewer 4: I believe the intent here was that `format` was referring to the proof format not the serialization to JSON or JSON-LD beforehand.

I tend to think just the single format property for proof format using the values @awoie has suggested from PE should suffice. I'd prefer to not complicate with both properties, but if the case can be made that this would simplify RP processing than I am open?

> `credential` : REQUIRED. A cryptographically verifiable proof in the defined proof format. 

Reviewer 3: ```suggestion
`credential` : REQUIRED. A W3C Verifiable Credential with a cryptographically verifiable proof in the defined proof format.
```

Reviewer 4: The question being if we want to limit to W3C VCs.. we have talked about support for broader `credentials`s that perhaps do not comply with the W3C spec too.

> Note that OP would first encode VCs using the rules defined in the Verifiable Credential specification either in JWT format or JSON-LD format, before encoded VCs as container objects.

Reviewer 3: ```suggestion
Note that OP would first encode VCs using the rules defined in the W3C Verifiable Credential specification either in JWT format or LD-Proof format, before encoded VCs as container objects.
```

Here is an example: 

```json
[
   {

>       "format":"w3cvc-jwt",

Reviewer 3: DIF PE would use ...
```suggestion
      "format":"jwt_vc",
```

      MzFjMjc2ZTEyZWNhYiNrZXlzLTEifQ.eyJzdWIiOiJkaWQ6ZXhhbXBsZTplYmZlYjFmNzEyZWJjNmYxY
      zI3NmUxMmVjMjEiLCJqdGkiOiJodHRwOi8vZXhhbXBsZS5lZHUvY3JlZGVudGlhbHMvMzczMiIsImlzc
      yI6Imh0dHBzOi8vZXhhbXBsZS5jb20va2V5cy9mb28uandrIiwibmJmIjoxNTQxNDkzNzI0LCJpYXQiO
      jE1NDE0OTM3MjQsImV4cCI6MTU3MzAyOTcyMywibm9uY2UiOiI2NjAhNjM0NUZTZXIiLCJ2YyI6eyJAY
      29udGV4dCI6WyJodHRwczovL3d3dy53My5vcmcvMjAxOC9jcmVkZW50aWFscy92MSIsImh0dHBzOi8vd
      3d3LnczLm9yZy8yMDE4L2NyZWRlbnRpYWxzL2V4YW1wbGVzL3YxIl0sInR5cGUiOlsiVmVyaWZpYWJsZ
      UNyZWRlbnRpYWwiLCJVbml2ZXJzaXR5RGVncmVlQ3JlZGVudGlhbCJdLCJjcmVkZW50aWFsU3ViamVjd
      CI6eyJkZWdyZWUiOnsidHlwZSI6IkJhY2hlbG9yRGVncmVlIiwibmFtZSI6IjxzcGFuIGxhbmc9J2ZyL
      UNBJz5CYWNjYWxhdXLDqWF0IGVuIG11c2lxdWVzIG51bcOpcmlxdWVzPC9zcGFuPiJ9fX19.KLJo5GAy
      BND3LDTn9H7FQokEsUEi8jKwXhGvoN3JtRa51xrNDgXDb0cq1UTYB-rK4Ft9YVmR1NI_ZOF8oGc_7wAp
      8PHbF2HaWodQIoOBxxT-4WNqAxft7ET6lkH-4S6Ux3rSGAmczMohEEf8eCeN-jC8WekdPl6zKZQj0YPB
      1rx6X0-xlFBs7cl6Wt8rfBP_tZ9YgVWrQmUWypSioc0MUyiphmyEbLZagTyPlUyflGlEdqrZAv6eSe6R
      txJy6M1-lD7a5HTzanYTWBPAUHDZGyGKXdJw-W_x0IWChBzI8t3kpG253fg6V3tPgHeKXE94fz_QpYfg
      --7kLsyBAfQGbg"
   }
]
```

### Verifiable Presentation

> A verifiable credential container is an array of objects, each of them containing the following fields:

Reviewer 3: ```suggestion
A verifiable presentation container is an array of objects, each of them containing the following fields:
```

> `format`: REQUIRED A JSON string denoting the proof format the presentation was returned in. This specification introduces the values `w3cvc-jsonld` and `w3cvc-jwt` to denote credentials in JSON-LD and JWT format, respectively.  

Reviewer 3: We might use the DIF PE denominators here as well ...
```suggestion
`format`: REQUIRED A JSON string denoting the proof format the presentation was returned in. This specification introduces the values `ldp_vp` and `jwt_vp` to denote credentials in LD-Proof and JWT format, respectively.  
```

> `presentation` : REQUIRED. A cryptographically verifiable proof in the defined proof format. 

Reviewer 3: ```suggestion
`presentation` : REQUIRED. A W3C Verifiable Presentation with a cryptographically verifiable proof in the defined proof format. 
```

> Note that OP would first encode VPs using the rules defined in the Verifiable Credential specification either in JWT format or JSON-LD format, before encoded VPs as container objects.

Reviewer 2: Is the assumption that VP in JWT contains VC in JWT and VP in LD-Proof contains VP in LD-Proof? If so, probably worth adding a note.

> Note that OP would first encode VPs using the rules defined in the Verifiable Credential specification either in JWT format or JSON-LD format, before encoded VPs as container objects.

Reviewer 3: ```suggestion
Note that OP would first encode VPs using the rules defined in the W3C Verifiable Credential specification either in JWT format or LD-Proof format, before encoded VPs as container objects.
```

Here is an example: 

```json
[
   {

>       "format":"w3cvc-ldp",

Reviewer 3: If DIF PE terminology adopted, then ...
```suggestion
      "format":"ldp_vp",
```

                  "type":"Ed25519Signature2018",
                  "created":"2021-03-19T15:30:15Z",
                  "jws":"eyJhbGciOiJFZERTQSIsImI2NCI6ZmFsc2UsImNyaXQiOlsiYjY0Il19..PT8yCqVjj5ZHD0W36zsBQ47oc3El07WGPWaLUuBTOT48IgKI5HDoiFUt9idChT_Zh5s8cF_2cSRWELuD8JQdBw",
                  "proofPurpose":"assertionMethod",
                  "verificationMethod":"did:example:issuer#keys-1"
               }
            }
         ],
         "id":"ebc6f1c2",
         "holder":"did:example:holder",
         "proof":{
            "type":"Ed25519Signature2018",
            "created":"2021-03-19T15:30:15Z",
            "challenge":"()&)()0__sdf",
            "jws":"eyJhbGciOiJFZERTQSIsImI2NCI6ZmFsc2UsImNyaXQiOlsiYjY0Il19..GF5Z6TamgNE8QjE3RbiDOj3n_t25_1K7NVWMUASe_OEzQV63GaKdu235MCS3hIYvepcNdQ_ZOKpGNCf0vIAoDA",
            "proofPurpose":"authentication",
            "verificationMethod":"did:example:holder#key-1"
         }
      }
   }
]
```

## JWT parameters extention

Verifiable credential objects can be exchanged between OP and RP enveloped in JWT claims in ID tokens or userinfo responses.  

This specification introduces the following JWT claims for that purpose:

> - `verifiable_credentials`:  A claim whose value is a verifiable credentials container object es defined above. 

Reviewer 2: *as defined above

> - `verifiable_presentations`:  A claim whose value is a verifiable presentations container object es defined above.

Reviewer 2: *as defined above

## Requesting W3C Verifiable Credential Objects 

This section illustrates how the `claims` parameter can be used for requesting verified presentations. It serves as a starting point to drive discussion about this aspect. There are other candidate approaches for this purpose (most notably [DIF Presentation Exchange](https://identity.foundation/presentation-exchange/). They will be evaluated as this draft evolves. 

### Requesting Verifiable Presentations

A Verifiable Presentation embedded in an ID Token (or userinfo response) is requested by adding an element `verifiable_presentations` to the `id_token` (or `userinfo`) top level element of the `claims` parameter. This element must contain the following element:

`credential_types`
Object array containing definitions of credential types the RP wants to obtain along with an (optional) definition of the claims from the respective credential type the RP is requesting. Each of those object has the following fields:

* `type` REQUIRED String denoting a credential type
* `claims` OPTIONAL An object determining the claims the RP wants to obtain using the same notation as used underneath `id_token`. 

> * `format` OPTION String designating the VP format. Predefined values are `w3cvp-jsonld` and `w3cvp-jwt`. 

Reviewer 3: ```suggestion
* `format` OPTION String designating the VP format. Predefined values are `ldp_vp` and `jwt_vp`. 
```

               "claims":{
                  "given_name":null,
                  "family_name":null,
                  "birthdate":null
               }
            }
         ]
      }
   }
}
```
#### Requesting Verifiable Credentials

A Verifiable Credential embedded in an ID Token (or userinfo response) is requested by adding a element `verifiable_credentials` to the `id_token` (or `userinfo`) top level element of the `claims` parameter. This element must contain a `credential_types` sub element as defined above.

### Requesting a VP Token

A VP Token is requested by adding a new top level element `vp_token` to the `claims` parameter. This element contains the sub elements as defined above.

### Requesting a VC Token

A VP Token is requested by adding a new top level element `vc_token` to the `claims` parameter. This element must contain a `credential_types` sub element as defined above.

##  Examples 

This section illustrates examples when W3C Verifiable Credentials objects are requested using `claims` parameter and returned inside ID Tokens.

### Self-Issued OpenID Provider with Verifiable Presentation in ID Token 

> Below are the examples when W3C Verifiable Credentials are requested and returned inside ID Token as part of Self-Issued OP response. ID Token contains a `verifiable_presentations` claim with the Verifiable Presentations data. It can also contain `verifiable_credentials` element with the Verifiable Credentials data. 

Reviewer 3: ```suggestion
Below are the examples when W3C Verifiable Credentials are requested and returned inside ID Token as part of Self-Issued OP response. ID Token contains a `verifiable_presentations` claim with the Verifiable Presentation data. It can also contain `verifiable_credentials` element with the Verifiable Credential data. 
```

    }   
  }
```
### Self-Issued OpenID Provider with Verifiable Presentation in ID Token (selective disclosure)
#### `claims` parameter 

Below is a non-normative example of how the `claims` parameter can be used for requesting verified presentations signed as Linked Data Proofs.

```json
{
   "id_token":{
      "verifiable_presentations":[
         {
            "credential_types":[
               "https://www.w3.org/2018/credentials/examples/v1/IDCardCredential"
            ],
            "claims":{
               "given_name":null,
               "family_name":null,
               "birthdate":null
            }
         }
      ]
   }
}
```

#### Authentication Response 

Below is a non-normative example of ID Token that includes `verifiable_presentations` claim.

```json
{
   "iss":"https://self-issued.me",
   "aud":"https://book.itsourweb.org:3000/client_api/authresp/uhn",
   "iat":1615910538,
   "exp":1615911138,
   "sub":"did:ion:EiC6Y9_aDaCsITlY06HId4seJjJ...b1df31ec42d0",
   "auth_time":1615910535,
   "verifiable_presentations":[
      {

>          "format":"w3cvp-jwt",

Reviewer 3: ```suggestion
         "format":"jwt_vp",
```

  &redirect_uri=https%3A%2F%2Fclient.example.org%2Fcb
```

### Authorization Code Flow with Verifiable Presentation returned from the UserInfo endpoint

> Below are the examples when verifiable presentation is requested and returned from the UserInfo endpoint as part of OpenID Connect Authorization Code Flow. UserInfo response contains a `verifiable_presentazions` element with the Verifiable Presentation data. It can also contain `verifiable_credentials` element with the Verifiable Credentials data. 

Reviewer 3: ```suggestion
Below are the examples when verifiable presentation is requested and returned from the UserInfo endpoint as part of OpenID Connect Authorization Code Flow. UserInfo response contains a `verifiable_presentations` element with the Verifiable Presentation data. It can also contain `verifiable_credentials` element with the Verifiable Credentials data. 
```

   "name": "Jane Doe",
   "given_name": "Jane",
   "family_name": "Doe",
    "verifiable_presentations":[
      {

>          "format":"w3cvp-jwt",

Reviewer 3: ```suggestion
         "format":"jwt_vp",
```

}
```

#### UserInfo Response 

Below is a non-normative example of a UserInfo Response that includes `verifiable_presentations` claim:

```json
  HTTP/1.1 200 OK
  Content-Type: application/json

  {
   "sub":"248289761001",
   "name":"Jane Doe",
   "given_name":"Jane",
   "family_name":"Doe",
   "verifiable_presentations":[
      {

>          "format":"w3cvp-jwt",

Reviewer 3: ```suggestion
         "format":"jwt_vp",
```

#### vp_token content

```json
[
   {

>       "format":"w3cvp-ldp",

Reviewer 3: ```suggestion
      "format":"ldp_vp",
```

  &redirect_uri=https%3A%2F%2Fclient.example.org%2Fcb
```

### Token Response (including vp_token)

```json
{
   "access_token":"SlAV32hkKG",
   "token_type":"Bearer",
   "refresh_token":"8xLOxBtZp8",
   "expires_in":3600,
   "id_token":"eyJ0 ... NiJ9.eyJ1c ... I6IjIifX0.DeWt4Qu ... ZXso",
   "vp_token":[
      {

>          "format":"w3cvp-ldp",

Reviewer 3: ```suggestion
         "format":"ldp_vp",
```
